src/modules/form/components/refactor/useFormDefinitionHandlers.tsx

```diff
@@ -0,0 +1,186 @@
+import { cloneDeep } from '@apollo/client/utilities';
+import { mapKeys, omit } from 'lodash-es';
+import { useCallback, useMemo } from 'react';
+import { FieldValues, useForm } from 'react-hook-form';
+import useComputedData from '../../hooks/useComputedData';
+import { LocalConstants } from '../../types';
+import {
+  addDescendants,
+  autofillValues,
+  dropUnderscorePrefixedKeys,
+  getInitialValues,
+  shouldEnableItem,
+} from '../../util/formUtil';
+import {
+  DisabledDisplay,
+  FormDefinitionJson,
+  FormItem,
+  ValidationError,
+} from '@/types/gqlTypes';
+
+export interface UseDynamicFormArgs<T extends FieldValues> {
+  definition: FormDefinitionJson;
+  initialValues?: T;
+  localConstants?: LocalConstants;
+  errors?: ValidationError[];
+}
+
+export const getSafeLinkId = (linkId: string) =>
+  String(linkId).replace(/\./g, ',');
+export const getCleanedLinkId = (linkId: string) =>
+  String(linkId).replace(/,/g, '.');
```

Comment:
I think the dots correspond to HUD spec, so I'm unsure how easy/okay it would be to replace them with another character at the backend or API level. I agree with the sentiment that the comma isn't ideal, and I'm open to the option of another replacement character. Though I think using any character that could possibly be part of a logical ID isn't the best, since it's possible it would be used in link IDs innocently and then cause issues or need special handling. For that reason I don't think using `-` or `_` is a good idea. My logic for using a comma is: 1) It reads as related to a period, so it's maybe more clear what character it's escaping; and 2) I would never think to use a comma in an ID, so I can't see one being used in a link ID in a future form definition.

In terms of picking a replacement for a comma, I see a couple of options:

1. Use some other character that wouldn't typically be included in IDs, maybe: `;`, `:`, `|`, `^`, `~`, `%`, etc.
2. Use an escape sequence of some kind: `:dot:` or some such.
3. Hash or base64 the link IDs. This would do it, and would probably make this a non-issue for any other reserved character issues, but probably not ideal since the internal form state would be hard to read.

What do you think?

Reply:
> For that reason I don't think using - or _ is a good idea

I imagine double or triple underscore would be safe (we'd leave single underscores alone and transform with `s/\./__/ and s/__/\./`).

Here's another idea:
These are HUD defined values so maybe we prefix them all with 'hud'? That would have the extra benefit of allowing us to target these properties with selectors (currently these ids that start with numeric chars are hard to use in js/css).

I'm thinking something like `hud-1-2-3` would be ideal, it looks nice and is not too hard to transform back to the original value with something like `/^hud-(\d+)-(\d+)-(\d+)/`. The 'hud' prefix should make it safe to use simple hyphens or underscores in place of dots.

Reply:
I changed it to `__`

Reply:
https://www.pivotaltracker.com/story/show/186922374